File: samples/sample.rb
#!/usr/bin/env ruby
# Ruby sample demonstrating all token types and syntax

# Single line comment
=begin
Multi-line comment
with multiple lines
=end

# Strings
string1 = "double quoted"
string2 = 'single quoted'
string3 = %q(percent q string)
string4 = %Q(percent Q with #{"interpolation"})
symbol = :symbol_name
regex = /pattern/i
regex2 = %r{alternative regex}i

# String interpolation
name = "World"
greeting = "Hello, #{name}!"
escaped = "escaped\nstring"
heredoc = <<-HEREDOC
  This is a
  multi-line heredoc
  string
HEREDOC

# Numbers
integer = 42
negative = -100
float = 3.14159
scientific = 1.5e-3
hex = 0xFF
octal = 0o755
binary = 0b1010
underscored = 1_000_000

# Booleans and nil
bool_true = true
bool_false = false
null_value = nil

# Arrays
array1 = [1, 2, 3, 4, 5]
array2 = ["a", "b", "c"]
array3 = %w(word1 word2 word3)
array_nested = [[1, 2], [3, 4]]
array_symbols = %i(symbol1 symbol2 symbol3)

# Hashes
hash1 = { "key" => "value", "name" => "John" }
hash2 = { key: "value", name: "John" }
hash3 = { "a" => 1, "b" => 2, "c" => 3 }
hash_nested = { person: { name: "John", age: 30 } }

# Ranges
range_inclusive = 1..10
range_exclusive = 1...10
range_string = "a".."z"

# Variables and constants
local_var = "local"
@instance_var = "instance"
@@class_var = "class"
$global_var = "global"
CONSTANT = "constant"

# String operations
str = "hello"
str_upper = str.upcase
str_reverse = str.reverse
str_length = str.length

# Array operations
arr = [1, 2, 3, 4, 5]
first = arr.first
last = arr.last
pushed = arr << 6
combined = arr + [7, 8]

# Hash operations
h = { a: 1, b: 2 }
value = h[:a]
keys = h.keys
values = h.values

# Functions (methods)
def simple_function
  puts "Hello"
end

def function_with_args(name, age)
  "#{name} is #{age} years old"
end

def function_with_defaults(name = "John", age = 30)
  { name: name, age: age }
end

def function_with_splat(*args)
  args
end

def function_with_double_splat(**kwargs)
  kwargs
end

def function_with_block
  yield if block_given?
end

def function_with_block_param(&block)
  block.call if block
end

# Return values
def get_value
  return 42
  # last expression is return value
  "not reached"
end

# Classes
class Animal
  # Class variables
  @@count = 0
  
  # Constructor
  def initialize(name, age = 0)
    @name = name
    @age = age
    @@count += 1
  end
  
  # Getter methods
  def name
    @name
  end
  
  def age
    @age
  end
  
  # Setter methods
  def name=(value)
    @name = value
  end
  
  # Attr shortcuts
  attr_reader :name, :age
  attr_writer :name
  attr_accessor :name, :age
  
  # Regular method
  def speak
    puts "#{@name} makes a sound"
  end
  
  # Method with ? (predicate)
  def adult?
    @age >= 18
  end
  
  # Method with ! (mutator)
  def grow!
    @age += 1
  end
  
  # Class method
  def self.count
    @@count
  end
  
  # Class method alternative
  class << self
    def create_default
      new("Unknown")
    end
  end
end

# Inheritance
class Dog < Animal
  def initialize(name, age, breed)
    super(name, age)
    @breed = breed
  end
  
  def speak
    puts "#{@name} says Woof!"
  end
  
  def breed
    @breed
  end
end

# Modules (mixins)
module Drawable
  def draw
    puts "Drawing #{self.class.name}"
  end
end

module Comparable
  def compare(other)
    self.age <=> other.age
  end
end

class Shape
  include Drawable
  include Comparable
end

# Singleton class
obj = Object.new
class << obj
  def singleton_method
    "Only on this object"
  end
end

# Control flow
def control_flow(x)
  # if-elsif-else
  if x > 0
    puts "positive"
  elsif x < 0
    puts "negative"
  else
    puts "zero"
  end
  
  # unless
  unless x.zero?
    puts "not zero"
  end
  
  # case-when
  case x
  when 1
    "one"
  when 2, 3
    "two or three"
  else
    "other"
  end
  
  # Modifier if
  puts "positive" if x > 0
  
  # Modifier unless
  puts "not zero" unless x.zero?
  
  # Modifier case
  result = case x
  when 1 then "one"
  when 2 then "two"
  else "other"
  end
  
  # Ternary
  message = x > 0 ? "positive" : "non-positive"
  
  # for loop
  for i in 1..5
    puts i
  end
  
  # while loop
  i = 0
  while i < 5
    puts i
    i += 1
  end
  
  # until loop
  i = 0
  until i >= 5
    puts i
    i += 1
  end
  
  # break and next
  [1, 2, 3, 4, 5].each do |num|
    next if num == 2
    break if num == 4
    puts num
  end
end

# Iterators and blocks
[1, 2, 3, 4, 5].each { |x| puts x }
[1, 2, 3].map { |x| x * 2 }
[1, 2, 3, 4].select { |x| x > 2 }
[1, 2, 3].reduce(0) { |sum, x| sum + x }

# Lambda and Proc
lambda_func = lambda { |x| x * 2 }
lambda_func2 = ->(x) { x * 2 }
proc_func = Proc.new { |x| x * 2 }

# Symbols
sym1 = :symbol
sym2 = :"symbol with spaces"
sym_from_string = "string".to_sym

# String manipulation
str = "hello world"
str_split = str.split(" ")
str_chars = str.chars
str_substitute = str.gsub("world", "ruby")

# Error handling
begin
  raise "Error message"
rescue => e
  puts e.message
rescue StandardError => e
  puts "Standard error: #{e}"
ensure
  puts "Cleanup"
end

# Operators
a = 10
b = 3

add = a + b
sub = a - b
mul = a * b
div = a / b
mod = a % b
power = a ** b

eq = a == b
neq = a != b
lt = a < b
gt = a > b
lte = a <= b
gte = a >= b
spaceship = a <=> b

and_op = true && false
or_op = true || false
not_op = !true

bit_and = a & b
bit_or = a | b
bit_xor = a ^ b
bit_not = ~a
left_shift = a << 1
right_shift = a >> 1

# String operators
concat = "hello" + " world"
repeat = "ha" * 3

# Regex operators
matches = "hello" =~ /ell/
no_match = "hello" !~ /xyz/

# Special operators
ternary = a > b ? "a is greater" : "b is greater"
null_coalesce = nil || "default"

# Comments with special syntax
# TODO: Complete this
# FIXME: Fix the bug
# NOTE: Important
# HACK: Quick fix

File: samples/sample.php
<?php
/**
 * PHP sample demonstrating all token types and syntax
 */

// Single line comment
# Bash-style comment
/* Multi-line
   comment */

// String types
$single_quote = 'single quoted string';
$double_quote = "double quoted string";
$string_var = "embedded $single_quote";
$string_escape = "line1\nline2\ttabbed";
$heredoc = <<<EOT
    This is a heredoc string
    It can span multiple lines
    And preserve formatting
    EOT;
$nowdoc = <<<'EOT'
    This is a nowdoc string
    It doesn't interpret variables
    Like $this variable
    EOT;

// Numeric types
$integer = 42;
$negative = -100;
$float = 3.14159;
$scientific = 1.5e-3;
$hex = 0xFF;
$binary = 0b1010;
$octal = 0755;

// Boolean
$bool_true = true;
$bool_false = false;

// Null
$null_value = null;

// Arrays
$indexed_array = [1, 2, 3, 4, 5];
$associative_array = ["a" => 1, "b" => 2, "c" => 3];
$nested_array = [
    "group1" => [1, 2, 3],
    "group2" => ["x" => "a", "y" => "b"],
];

// Array functions
$array = [3, 1, 4, 1, 5, 9];
count($array);
array_push($array, 2);
array_pop($array);
array_shift($array);
array_unshift($array, 0);
in_array(3, $array);
array_key_exists("key", $associative_array);

// String functions
$str = "Hello World";
strlen($str);
strtoupper($str);
strtolower($str);
substr($str, 0, 5);
strpos($str, "World");
str_replace("World", "PHP", $str);
explode(" ", $str);
implode("-", ["a", "b", "c"]);
trim($str);

// Functions
function simpleFunction() {
    return "result";
}

function functionWithArgs($name, $age) {
    return "$name is $age years old";
}

function functionWithDefault($name = "John", $age = 30) {
    return [$name, $age];
}

function functionWithVariadic(...$args) {
    return count($args);
}

function &functionByReference() {
    global $globalVar;
    return $globalVar;
}

function functionWithTyping(int $x, string $y): string {
    return "$x: $y";
}

function functionWithNullable(?string $value): ?string {
    return $value;
}

function functionWithUnionType(int|float $number): string {
    return "Number: $number";
}

// Arrow functions (PHP 7.4+)
$arrow = fn($x) => $x * 2;

// Closures
$closure = function($x) use ($globalVar) {
    return $x + $globalVar;
};

// Variable scope
$global_var = "global";

function scopeExample() {
    global $global_var;
    $global_var;
    
    static $static_var = 0;
    $static_var++;
}

// Constants
define("MY_CONSTANT", "constant value");
const CLASS_CONSTANT = "class constant";

// Classes
class Animal {
    // Properties
    public $name;
    protected $age;
    private $secret;
    public static $count = 0;
    
    // Type hints
    public string $typed_property = "value";
    private int $age_typed;
    
    // Constructor
    public function __construct($name, $age = 0) {
        $this->name = $name;
        $this->age = $age;
        self::$count++;
    }
    
    // Regular method
    public function speak() {
        return "{$this->name} makes a sound";
    }
    
    // Protected method
    protected function protectedMethod() {
        return "protected";
    }
    
    // Private method
    private function privateMethod() {
        return "private";
    }
    
    // Static method
    public static function createDefault() {
        return new self("Unknown");
    }
    
    // Magic methods
    public function __get($name) {
        return $this->$name ?? null;
    }
    
    public function __set($name, $value) {
        $this->$name = $value;
    }
    
    public function __call($name, $args) {
        return "Called method $name";
    }
    
    public static function __callStatic($name, $args) {
        return "Called static method $name";
    }
    
    public function __toString() {
        return $this->name;
    }
    
    public function __invoke() {
        return "Invoked";
    }
    
    // Getters and setters
    public function getName() {
        return $this->name;
    }
    
    public function setName($name) {
        $this->name = $name;
    }
}

// Inheritance
class Dog extends Animal {
    private $breed;
    
    public function __construct($name, $breed) {
        parent::__construct($name);
        $this->breed = $breed;
    }
    
    public function speak() {
        return "{$this->name} says Woof!";
    }
    
    public function getBreed() {
        return $this->breed;
    }
}

// Interfaces
interface Reader {
    public function read();
}

interface Writer {
    public function write($data);
}

// Implementing interfaces
class FileHandler implements Reader, Writer {
    private $file;
    
    public function read() {
        return "reading";
    }
    
    public function write($data) {
        return "writing";
    }
}

// Abstract classes
abstract class Shape {
    protected $color;
    
    abstract public function area();
    abstract public function perimeter();
    
    public function getColor() {
        return $this->color;
    }
}

class Circle extends Shape {
    private $radius;
    
    public function __construct($radius) {
        $this->radius = $radius;
    }
    
    public function area() {
        return 3.14 * $this->radius * $this->radius;
    }
    
    public function perimeter() {
        return 2 * 3.14 * $this->radius;
    }
}

// Traits
trait Logger {
    public function log($message) {
        echo "Log: $message\n";
    }
}

trait Timestamp {
    public function getTimestamp() {
        return time();
    }
}

class Application {
    use Logger, Timestamp;
}

// Anonymous classes (PHP 7+)
$obj = new class {
    public function method() {
        return "anonymous class";
    }
};

// Control flow
function controlFlow($x) {
    // if-else
    if ($x > 0) {
        return "positive";
    } elseif ($x < 0) {
        return "negative";
    } else {
        return "zero";
    }
    
    // switch
    switch ($x) {
        case 1:
            return "one";
        case 2:
        case 3:
            return "two or three";
            break;
        default:
            return "other";
    }
    
    // Ternary
    $result = $x > 0 ? "positive" : "non-positive";
    
    // Null coalesce
    $value = $x ?? 0;
    
    // Spaceship operator
    $comparison = $x <=> 0;
    
    // for loop
    for ($i = 0; $i < 10; $i++) {
        if ($i == 3) continue;
        if ($i == 7) break;
    }
    
    // foreach loop
    foreach ([1, 2, 3] as $item) {
        echo $item;
    }
    
    foreach (["a" => 1, "b" => 2] as $key => $value) {
        echo "$key: $value";
    }
    
    // while loop
    while ($x < 10) {
        $x++;
    }
    
    // do-while loop
    do {
        $x++;
    } while ($x < 20);
}

// Exception handling
try {
    throw new Exception("Something went wrong");
} catch (Exception $e) {
    echo $e->getMessage();
} catch (Throwable $e) {
    echo "Throwable";
} finally {
    echo "Cleanup";
}

// Operators
$a = 10;
$b = 3;

// Arithmetic
$add = $a + $b;
$sub = $a - $b;
$mul = $a * $b;
$div = $a / $b;
$mod = $a % $b;
$pow = $a ** $b;

// String concatenation
$concat = "Hello" . " " . "World";

// Comparison
$eq = $a == $b;
$identical = $a === $b;
$neq = $a != $b;
$not_identical = $a !== $b;

// Logical
$and = true && false;
$or = true || false;
$xor = true xor false;
$not = !true;

// Bitwise
$band = $a & $b;
$bor = $a | $b;
$bxor = $a ^ $b;
$bnot = ~$a;
$shl = $a << 1;
$shr = $a >> 1;

// Assignment operators
$a += 5;
$b -= 2;
$a *= 2;
$b /= 2;
$a %= 3;
$a .= "string";

// instanceof
if ($obj instanceof Dog) {
    echo "Is a Dog";
}

// Comments with special syntax
// TODO: Implement this feature
// FIXME: Fix the bug
// NOTE: Important information
// HACK: Quick workaround

?>
